### PR TITLE
Latest Java17 NMT fix proposal

### PR DESCRIPTION
**Pull request by dkaukov**
In latest Java some NMT categories can disappear from set. I've noticed this behavior with "Tracing" category.  I suggest slightly simplify process:
* Fetch NMT periodically instead of rate limiting
* Cache gauge values instead set of categories, so they can keep last reported value

### CODE REVIEW

File: otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java [hunk 1]
```diff
@@ -14,128 +14,72 @@
...
-    private static final Logger LOG = LoggerFactory.getLogger(NmtMetrics.class);
     private static final Duration REFRESH_PERIOD = Duration.ofSeconds(10);
 
     private final String prefix;
     private final MetricRegistry metrics;
-    private final Set<String> registeredMetrics = new HashSet<>();
 
-    private Nmt nmt;
-    private Instant lastRefresh;
+    private final ScheduledExecutorService exec = Executors.newScheduledThreadPool(
```
Comment (mikebell90) [inline otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java:44]: This isn't set to set daemon = true so a shutdown issue is likely

File: otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java [hunk 2]
```diff
@@ -14,128 +14,72 @@
...
     }
 
-    private void refreshCategories() {
-        final Set<String> atStart = ImmutableSet.copyOf(registeredMetrics);
-        nmt.categories.keySet().forEach(categoryName -> {
-            final String metricName = categoryName.toLowerCase(Locale.ROOT).replaceAll(" ", "-");
-            if (registeredMetrics.contains(metricName)) {
+    private void setGaugeValue(String metricName, long value) {
+        AtomicLong gauge = this.gauges.get(metricName);
+        if (gauge == null) {
```
Comment (mikebell90) [inline otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java:64]: I guess I'm not completely convinced double reference check is that important here, but no matter

File: otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java [hunk 3]
```diff
@@ -14,128 +14,72 @@
...
     private final String prefix;
     private final MetricRegistry metrics;
-    private final Set<String> registeredMetrics = new HashSet<>();
 
-    private Nmt nmt;
-    private Instant lastRefresh;
+    private final ScheduledExecutorService exec = Executors.newScheduledThreadPool(
+        1,
+        new ThreadFactoryBuilder()
+            .setNameFormat("nmt-fetcher-%d")
```
Comment (mikebell90) [inline otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java:47]:
```suggestion
            .setNameFormat("nmt-fetcher-%d")
            .setDaemon(true)
```

File: otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java [hunk 4]
```diff
@@ -14,128 +14,72 @@
 package com.opentable.metrics.jvm;
 
 import java.time.Duration;
-import java.time.Instant;
```
[the remark this answers is not in the document]
Comment (dkaukov) [reply otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java:17]: no

File: otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java [hunk 5]
```diff
@@ -14,128 +14,72 @@
...
-        nmt.categories.keySet().forEach(categoryName -> {
-            final String metricName = categoryName.toLowerCase(Locale.ROOT).replaceAll(" ", "-");
-            if (registeredMetrics.contains(metricName)) {
+    private void setGaugeValue(String metricName, long value) {
+        AtomicLong gauge = this.gauges.get(metricName);
+        if (gauge == null) {
+            AtomicLong newGauge = new AtomicLong(value);
+            gauge = this.gauges.putIfAbsent(metricName, newGauge);
+            if (gauge == null) {
+                metrics.register(metricName, (Gauge<Long>) newGauge::longValue);
```
[the remark this answers is not in the document]
Comment (dkaukov) [reply otj-metrics-core/src/main/java/com/opentable/metrics/jvm/NmtMetrics.java:68]: but it is constructed with value `new AtomicLong(value);`

**Review by mikebell90 — APPROVED** [review]

**Pull request merged**